src/v/datalake/translation/deps.cc

```diff
@@ -52,26 +52,29 @@ ss::future<> writer_reservations_impl::maybe_reserve_memory(
     while (_available_memory < bytes) {
         auto reservation = co_await _reservations_tracker.reserve_memory(as);
         _available_memory += reservation.count();
-        _total_reserved_memory += reservation.count();
-        _reservations.push_back(std::move(reservation));
+        if (_reservations.count()) {
```

Comment:
nit: is it possible to simply call adopt if `_reservations` is empty ?

Reply:
In this case don't think we can.

adopt() checks if the adopting units belong to the same semaphore as existing  units (so they are all returned to the correct semaphore).  We have to have the backing semaphore address to call the appropriate constructor 

```
semaphore_units(basic_semaphore<ExceptionFactory, Clock>* sem, size_t n) noexcept : _sem(sem), _n(n) {}
```

or seed it with some units so the move constructor sets it (like in this case as semaphore is hidden in the implementation details).